### PR TITLE
[Merged by Bors] - chore: adaptations for nightly-2024-11-14

### PR DESCRIPTION
**Pull request by kim-em**

### CODE REVIEW

File: Mathlib/Algebra/Category/Ring/Epi.lean [hunk 1]
```diff
@@ -24,7 +24,7 @@ lemma CommRingCat.epi_iff_tmul_eq_tmul {R S : Type u} [CommRing R] [CommRing S]
   constructor
   · intro H
     have := H.1 (CommRingCat.ofHom <| Algebra.TensorProduct.includeLeftRingHom (R := R))
-      (CommRingCat.ofHom <| Algebra.TensorProduct.includeRight.toRingHom)
+      (CommRingCat.ofHom <| (Algebra.TensorProduct.includeRight (R := R) (A := S)).toRingHom)
```
Comment (jcommelin) [inline Mathlib/Algebra/Category/Ring/Epi.lean:27]: This makes me a little bit sad, but we can certainly live with this.

File: Mathlib/Algebra/Lie/Semisimple/Basic.lean [hunk 2]
```diff
@@ -212,7 +212,7 @@ lemma finitelyAtomistic : ∀ s : Finset (LieIdeal R L), ↑s ⊆ {I : LieIdeal
   set K := s'.sup id
   suffices I ≤ K by
     obtain ⟨t, hts', htI⟩ := finitelyAtomistic s' hs'S I this
-    exact ⟨t, hts'.trans hs'.subset, htI⟩
+    exact ⟨t, Finset.Subset.trans hts' hs'.subset, htI⟩
```
Comment (jcommelin) [inline Mathlib/Algebra/Lie/Semisimple/Basic.lean:215]: Do we understand this regression? This makes me moderately sad...

File: Mathlib/Data/DFinsupp/Sigma.lean [hunk 3]
```diff
@@ -63,14 +63,14 @@ theorem sigmaCurry_zero [∀ i j, Zero (δ i j)] :
 
 @[simp]
 theorem sigmaCurry_add [∀ i j, AddZeroClass (δ i j)] (f g : Π₀ (i : Σ _, _), δ i.1 i.2) :
-    sigmaCurry (f + g) = sigmaCurry f + sigmaCurry g := by
+    sigmaCurry (f + g) = (sigmaCurry f + sigmaCurry g : Π₀ (i) (j), δ i j) := by
```
Comment (jcommelin) [inline Mathlib/Data/DFinsupp/Sigma.lean:66]: This makes me a little bit sad, but I guess it's not too bad.

File: Mathlib/FieldTheory/Relrank.lean [hunk 4]
```diff
@@ -35,11 +35,13 @@ variable {E : Type v} [Field E] {L : Type w} [Field L]
 
 variable (A B C : Subfield E)
 
+set_option synthInstance.maxHeartbeats 400000 in
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Relrank.lean:38]: Can we add some adaptation notes here? Or is this just a fact of life that we have to live with until we debug/speedup?

File: Mathlib/MeasureTheory/Group/Measure.lean [hunk 5]
```diff
@@ -764,7 +764,7 @@ nonrec theorem _root_.MulEquiv.isHaarMeasure_map [BorelSpace G] [TopologicalGrou
     [TopologicalGroup H] (e : G ≃* H) (he : Continuous e) (hesymm : Continuous e.symm) :
     IsHaarMeasure (Measure.map e μ) :=
   let f : G ≃ₜ H := .mk e
-  isHaarMeasure_map μ e he e.surjective f.isClosedEmbedding.tendsto_cocompact
+  isHaarMeasure_map μ e.toMonoidHom he e.surjective f.isClosedEmbedding.tendsto_cocompact
```
Comment (jcommelin) [inline Mathlib/MeasureTheory/Group/Measure.lean:767]: Why do we need this now? Makes me mildly sad.

File: Mathlib/MeasureTheory/MeasurableSpace/Embedding.lean [hunk 6]
```diff
@@ -507,8 +507,10 @@ def arrowProdEquivProdArrow (α β γ : Type*) [MeasurableSpace α] [MeasurableS
   measurable_toFun _ h := by
     simp_rw [Equiv.arrowProdEquivProdArrow, coe_fn_mk]
     exact MeasurableSet.preimage h (Measurable.prod_mk
-        (measurable_pi_lambda (fun a c ↦ (a c).1) fun a ↦ (measurable_pi_apply a).fst)
-        (measurable_pi_lambda (fun a c ↦ (a c).2) fun a ↦ (measurable_pi_apply a).snd ))
+        (measurable_pi_lambda (fun (a : γ → α × β) c ↦ (a c).1)
```
Comment (jcommelin) [inline Mathlib/MeasureTheory/MeasurableSpace/Embedding.lean:510]: This becomes quite a bit more unreadable... mildly sad...

File: Mathlib/RingTheory/Jacobson.lean [hunk 7]
```diff
@@ -704,7 +704,7 @@ lemma finite_of_finite_type_of_isJacobsonRing (R S : Type*) [CommRing R] [Field
   obtain ⟨ι, hι, f, hf⟩ := Algebra.FiniteType.iff_quotient_mvPolynomial'.mp ‹_›
   have : (algebraMap R S).IsIntegral := by
     rw [← f.comp_algebraMap]
-    exact MvPolynomial.comp_C_integral_of_surjective_of_isJacobsonRing f hf
+    exact MvPolynomial.comp_C_integral_of_surjective_of_isJacobsonRing f.toRingHom hf
```
Comment (jcommelin) [inline Mathlib/RingTheory/Jacobson.lean:707]: Why do we need this now?

**Review by jcommelin — COMMENTED** [review]
I did a stream of consciousness review, leaving comments where I got sad enough that it was worth commenting about.

I think it's worth discussing these a bit more, and possibly adding adaptation notes recording our conclusions.

**Comment by kbuzzard** [comment]
These changes are sometimes regressions, but it's not like we don't have plenty of other occurrences of `.toRingHom` in mathlib already (I mean, literally hundreds). 

bors d+

**Comment by kim-em** [comment]
Nearly all of the regressions are due to a horrible bug that Kyle discovered and fixed in https://github.com/leanprover/lean4/pull/6024. Strangely, this bug was actually a *helpful* bug, that was conveying type hints into some unification problems, but via a really terrible mechanism. Kyle is thinking about ways we could improve the situation, but the squashing the bug is essential.

**Comment by kim-em** [comment]
I've added adaptation notes for all of the regressions Johan noted above. (Note there were already many backported to master in #18896, but I think these all had good notes already.)

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**